aten/src/ATen/native/Copy.cpp

```diff
@@ -42,6 +42,10 @@ Tensor& _s_copy__cpu(Tensor& self, const Tensor& src, bool non_blocking) {
     _s_copy_from(src, self, non_blocking);
     return self;
   }
+  if (src.is_xla()) {
+    self = src.type().to(src, self, non_blocking, true);
+    return self;
```

Comment:
I think we're going to need an explanation here why this special case is necessary. Naively, I would have expected changing the line above `src.is_cuda()` to `!src.is_cpu()` would have been sufficient to make the double dispatch work out.

Reply:
I guess I was confused by the s_copy_from signature, taking a const Tensor& and copying into it.
I though the function was supposed to create a new tensor of type(self), with the data coming from dst.
We can special case s_copy_from from our auto generated code and handle it that way.

Reply:
There is no is_cpu() in Tensor. I can so is_cuda||is_xla I guess.
But what bother me more is a signature taking a const&, and writing values into it.

Reply:
> There is no is_cpu() in Tensor. I can so is_cuda||is_xla I guess.

Let's just add an `is_cpu`.  I can do it.

> But what bother me more is a signature taking a const&, and writing values into it.

To quote some of @ezyang's writing on the topic:

Tensor is [...] a pointer type. Think of it as being equivalent to shared_ptr<TensorImpl>. If you are a native C++ programmer, this behavior is unusual; fortunately, we expect this to be the only visible C++ class in PyTorch 1.0 which has this behavior.

**Corollary:** Given that Tensor is a pointer already, this means Tensor* should be avoided. There are a number of legacy APIs which are spelled this way (for example, Operator::Output), but new code should avoid taking a pointer to Tensor. const Tensor&, on the other hand, is still idiomatic, because it avoids reference count bumps and at use-sites is interchangeable with Tensor.

**Corollary:** const Tensor& doesn't mean “immutable tensor”, you can still mutate a tensor with this type; const Tensor& means you have a const reference to a mutable tensor. (A lot of people find this surprising, but it’s an unavoidable consequence of making Tensor a pointer type.)